File: mementoPattern/csharp/Originator.cs
class Originator
{
	private string _state;

	public Originator(string state)
	{
		_state = state;
	}

	public Memento CreateSnapshot()
	{
		System.Console.WriteLine("Originator: Create snapshot");
		return new Memento(_state);
	}

	public void LoadSnapshot(Memento snapshot)
	{
		System.Console.WriteLine("Originator: Load snapshot");
		_state = snapshot.GetState();
	}

	public void BusinessOperation()
	{
		System.Console.WriteLine("Originator: Perform business operation");
		_state += ("+");
	}
}

File: mementoPattern/originator.cpp
#include "./originator.h"
#include "./memento.h"

Memento *Originator::CreateSnapshot()
{
	printf("Originator: Create snapshot\n");
	return new Memento(_state);
}

void Originator::LoadSnapshot(Memento *snapshot)
{
	printf("Originator: Load snapshot\n");
	_state = snapshot->GetState();
}

void Originator::BusinessOperation()
{
	printf("Originator: Perform business operation\n");
	_state.append("+");
}